pkg/controller/ttl/ttlcontroller.go

```diff
@@ -0,0 +1,299 @@
+	ttlc.enqueueNode(node)
+}
+
+func (ttlc *TTLController) updateNode(_, newObj interface{}) {
+	node, ok := newObj.(*v1.Node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", newObj))
+		return
+	}
+	// Processing all updates of nodes guarantees that we will update
+	// the ttl annotation, when cluster size changes.
+	// We are relying on the fact that Kubelet is updating node status
+	// every 10s (or generaly every X seconds), which means that whenever
+	// required, its ttl annotation should be updated within that period.
+	ttlc.enqueueNode(node)
+}
+
+func (ttlc *TTLController) deleteNode(obj interface{}) {
+	_, ok := obj.(*v1.Node)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", obj))
+			return
+		}
+		_, ok = tombstone.Obj.(*v1.Node)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("unexpected object types: %v", obj))
+			return
+		}
+	}
+	atomic.AddInt32(&ttlc.nodeCount, -1)
+	// We are not processing the node, as it no longer exists.
+}
+
+func (ttlc *TTLController) enqueueNode(node *v1.Node) {
+	key, err := controller.KeyFunc(node)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v", node)
+		return
+	}
+	ttlc.queue.Add(key)
+}
+
+func (ttlc *TTLController) worker() {
+	for ttlc.processItem() {
+		if quit := ttlc.processItem(); quit {
+			glog.Infof("TTL controller worker shutting down")
+			return
+		}
+	}
+}
+
+func (ttlc *TTLController) processItem() bool {
+	key, quit := ttlc.queue.Get()
+	if quit {
+		return false
+	}
+	defer ttlc.queue.Done(key)
+
+	err := ttlc.updateNodeIfNeeded(key.(string))
+	if err == nil {
+		ttlc.queue.Forget(key)
+		return true
+	}
+
+	ttlc.queue.AddRateLimited(key)
+	utilruntime.HandleError(err)
+	return true
+}
+
+type ttlBoundary struct {
+	nodeCount  int32
+	ttlSeconds int
+}
+
+var (
+	ttlBoundaries = []ttlBoundary{
+		{nodeCount: 100, ttlSeconds: 0},
+		{nodeCount: 500, ttlSeconds: 15},
+		{nodeCount: 1000, ttlSeconds: 30},
+		{nodeCount: 2000, ttlSeconds: 60},
+		{nodeCount: 5000, ttlSeconds: 300},
+	}
+)
+
+func (ttlc *TTLController) getDesiredTTLSeconds(currentTTL int) int {
+	nodeCount := atomic.LoadInt32(&ttlc.nodeCount)
+	// To avoid situation when we are changing the annotation all the time,
+	// because the cluster size is flapping between X and X+1 nodes (e.g.
+	// due to cluster autoscaler), we use the following algorithm:
+	// - we compute the desired ttl first based on current cluster size.
+	// - if we are supposed to lower the current ttl, the current cluster
+	//   size must be at most boundary -1
+	// Note that with this algorith, to have flapping ttls on nodes, the cluster
+	// has to be flapping constantly betwenn <=X-1 and X+1<=.
+	// WARNING: you can't make this "1" a higher constant, as it may result
+	// in different nodes having different ttls over long period of time, e.g.
+	// if we swap the 1 with say 5, then in the scenario:
```

Comment:
Not sure why this is a problem. You can modify the internal, desired TTL whenever the node count changes. All nodes can just use this internal TTL (instead of calculating it individually). Basically, only change the TTL if the N (the node count) is greater than Tmax or fall below Tmin, where Tmax - Tmin > M. M can be a buffer that's larger than 1.

Having a buffer of only 1 (the current implementation) doesn't seem safe, especially in a 5000 node cluster.